File: QualificationRound/cd.py
try:
	import os, sys
	stdin = sys.stdin
	# stdin = open('s.txt')
	stdin = open(os.path.splitext(__file__)[0] + '.txt')
	input = lambda: stdin.readline()[:-1]
except Exception:
	pass


import math, operator
T = int(input())
P = int(input())

def f(x):
	return 1 / (1 + math.e**(-x))

for test in range(T):
	result = []
	for i in range(100):
		result.append(input())
	skill = []
	for i in result:
		correct_rate = i.count('1') / 10000
		# f(x) := 1 / (1 + %e**(-x));
		# plot2d(integrate(f(s - q), q, -3, 3) / 6, [s, -3, 3]);
		# plot2d(((integrate(f(s - q), q, -3, 3) / 6) - 0.5) * 7.5, [s, -3, 3]);
		skill.append((correct_rate - 0.5) * 7.5)
	diff = []
	for i in range(10000):
		guess = [-2.75, -2.25, -1.75, -1.25, -0.75, -0.25, 0.25, 0.75, 1.25,
					1.75, 2.25, 2.75]
		prob = [0] * len(guess)
		for r, s in zip(result, skill):
			correct = r[i] == '1'
			for index, q in enumerate(guess):
				if correct:
					prob[index] += f(s - q)
				else:
					prob[index] += 1 - f(s - q)
		diff.append(guess[max(enumerate(prob), key=operator.itemgetter(1))[0]])
	fukashigis = []
	for i in range(100):
		fukashigi = 0
		for j in range(10000):
			if result[i][j] == '1':
				fukashigi += 1 - f(skill[i] - diff[j])
			else:
				fukashigi += f(skill[i] - diff[j])
		fukashigis.append(fukashigi)
	print(skill)
	print(diff)
	print(fukashigis)
	ans = min(enumerate(fukashigis), key=operator.itemgetter(1))[0] + 1
	print('Case #%d:' % (test + 1), ans)

File: QualificationRound/cd.cpp
#include <iostream>
#include <cmath>
#include <cassert>

using namespace std;

#define FLOAT float
#define EXP expf

FLOAT f(FLOAT x) {
	return 1 / (1 + EXP(-x));
}

FLOAT GUESS[] = {-2.75, -2.25, -1.75, -1.25, -0.75, -0.25, 0.25, 0.75, 1.25,
					1.75, 2.25, 2.75};
constexpr int LEN_GUESS = sizeof(GUESS) / sizeof(GUESS[0]);

template <typename T>
T square(T x) {
	return x * x;
}

int main() {
	int T, P;
	cin >> T >> P;
	for (int test = 0; test < T; test++) {
		char result[100][10000];
		for (int i = 0; i < 100; i++)
			cin >> result[i];
		
//		for (int i = 0; i < 100; i++)
//			for (int j = 0; j < 10000; j++)
//				assert (result[i][j] == '1' || result[i][j] == '0');
		FLOAT skill[100];
		for (int i = 0; i < 100; i++) {
			int correct_count = 0;
			for (int j = 0; j < 10000; j++)
				correct_count += result[i][j] == '1' ? 1 : 0;
			FLOAT correct_rate = ((FLOAT) correct_count) / 10000;
			skill[i] = (correct_rate - 0.5) * 7.5;
		}
		FLOAT diff[10000];
		for (int i = 0; i < 10000; i++) {
			FLOAT prob[LEN_GUESS];
			for (int j = 0; j < LEN_GUESS; j++)
				prob[j] = 0;
			for (int j = 0; j < 100; j++)
				for (int k = 0; k < LEN_GUESS; k++) {
					if (result[j][i] == '1')
						prob[k] += f(skill[j] - GUESS[k]);
					else
						prob[k] += 1 - f(skill[j] - GUESS[k]);
				}
			int correct_guess = 0;
			for (int j = 0; j < LEN_GUESS; j++)
				if (prob[correct_guess] < prob[j])
					correct_guess = j;
			diff[i] = GUESS[correct_guess];
		}
		FLOAT fukashigis[100];
		for (int i = 0; i < 100; i++) {
			FLOAT fukashigi = 0;
			if (0) {
				for (int j = 0; j < 10000; j++) {
					if (result[i][j] == '1')
						fukashigi -= 1 - f(skill[i] - diff[j]);
					else
						fukashigi -= f(skill[i] - diff[j]);
				}
			} else if (0) {
				for (int j = 0; j < 10000; j++) {
					fukashigi += result[i][j] - '0';
					fukashigi -= f(skill[i] - diff[j]);
				}
			} else {
				for (int j = 0; j < 10000; j++) {
					if (result[i][j] == '1')
						fukashigi -= square(1 - f(skill[i] - diff[j]));
					else
						fukashigi -= square(f(skill[i] - diff[j]));
				}
			}
			fukashigis[i] = fukashigi;
		}
		int ans = 0;
		for (int i = 0; i < 100; i++) {
			if (fukashigis[ans] < fukashigis[i])
				ans = i;
		}
		cout << "Case #" << (test + 1) << ": " << (ans + 1) << endl;
	}
}

